[Quant] Use finite hops to check if the quant nodes are connected with producer

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108572

Summary:
Using dfs to check if two nodes are connecgted is making it very slow. Even BFS was the same. So resorted to finite hop checks starting from dq node to find a node that is not inserted by quant workflow.

Test Plan:
https://gist.github.com/leslie-fang-intel/9cd828623f567a3afbf41564d3546398

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48971710](https://our.internmc.facebook.com/intern/diff/D48971710)

kimishpatel committed:
[Quant] Move to BFS instead of DFS to check for connectedness

Summary:
Using dfs to check if two nodes are connecgted is making it very slow.
Use of BFS makes it much faster.

Test Plan:
https://gist.github.com/leslie-fang-intel/9cd828623f567a3afbf41564d3546398

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48971710).

jerryzh168 commented:
thanks, please add a comment/doc block to talk about the cleanup plans as well so that we don't forget

kimishpatel committed:
Update on "[Quant] Move to BFS instead of DFS to check for connectedness"


Summary:
Using dfs to check if two nodes are connecgted is making it very slow.
Use of BFS makes it much faster.

Test Plan:
https://gist.github.com/leslie-fang-intel/9cd828623f567a3afbf41564d3546398

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48971710](https://our.internmc.facebook.com/intern/diff/D48971710)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48971710).

kimishpatel committed:
Update on "[Quant] Move to BFS instead of DFS to check for connectedness"


Summary:
Using dfs to check if two nodes are connecgted is making it very slow.
Use of BFS makes it much faster.

Test Plan:
https://gist.github.com/leslie-fang-intel/9cd828623f567a3afbf41564d3546398

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48971710](https://our.internmc.facebook.com/intern/diff/D48971710)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48971710).

leslie-fang-intel commented:
Thanks for the fix.

leslie-fang-intel commented:
oh... seems still has issue when running resnet152: https://gist.github.com/leslie-fang-intel/5783540cd11e9c132630319bde16a422

leslie-fang-intel reviewed (changes requested):
oh... seems still has issue when running resnet152: https://gist.github.com/leslie-fang-intel/5783540cd11e9c132630319bde16a422

jerryzh168 commented:
@kimishpatel maybe we have to go with traversing the hardcoded q/dq path

kimishpatel committed:
Update on "[Quant] Move to BFS instead of DFS to check for connectedness"


Summary:
Using dfs to check if two nodes are connecgted is making it very slow.
Use of BFS makes it much faster.

Test Plan:
https://gist.github.com/leslie-fang-intel/9cd828623f567a3afbf41564d3546398

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48971710](https://our.internmc.facebook.com/intern/diff/D48971710)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48971710).

kimishpatel commented:
> @kimishpatel maybe we have to go with traversing the hardcoded q/dq path

ok this is a step closer to hardcoding but not quite hardcoding. @jerryzh168 how do you try this on inception v4? can you copy paste command here?

jerryzh168 commented on torch/ao/quantization/pt2e/utils.py:
```diff
@@ -34,15 +35,17 @@
+    finds if source and dest are connected. Assumption is that only quant workflow
+    inserted ops exist between source and dest
+    """
+    quant_workflow_ops = _QUANTIZE_OPS + _DEQUANTIZE_OPS
```
maybe add choose_qparams op as well for dynamic quant?

kimishpatel replied:
its there in the next line

jerryzh168 commented:
> > @kimishpatel maybe we have to go with traversing the hardcoded q/dq path
> 
> ok this is a step closer to hardcoding but not quite hardcoding. @jerryzh168 how do you try this on inception v4? can you copy paste command here?

I have a diff, could you import to phabricator? I can rebase and try this out

kimishpatel commented:
> this

it is imported to phabricator at [D48971710](https://our.internmc.facebook.com/intern/diff/D48971710)

kimishpatel committed:
Update on "[Quant] Move to BFS instead of DFS to check for connectedness"


Summary:
Using dfs to check if two nodes are connecgted is making it very slow.
Use of BFS makes it much faster.

Test Plan:
https://gist.github.com/leslie-fang-intel/9cd828623f567a3afbf41564d3546398

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D48971710](https://our.internmc.facebook.com/intern/diff/D48971710)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48971710).
